annotation/key.go

```diff
@@ -62,6 +63,85 @@ func (k FieldAnnotationKey) String() string {
 	return fmt.Sprintf("Field %s", k.FieldDecl.Name())
 }
 
+// ArgAnnotationKey is similar to ParamAnnotationKey but it represents the site in the caller where
+// the actual argument is passed to the called function. For the same parameter of the same
+// function, there is only one distinct ParamAnnotationKey but there is a new ArgAnnotationKey for
+// the parameter for every call of the same function.
+type ArgAnnotationKey struct {
+	FuncDecl *types.Func
+	ParamNum int
+	Location token.Position
+}
+
+// ParamName returns the *types.Var naming the parameter associate with this key
+// nilable(result 0)
+func (ak ArgAnnotationKey) ParamName() *types.Var {
+	return ak.FuncDecl.Type().(*types.Signature).Params().At(ak.ParamNum)
+}
+
+// Lookup looks this key up in the passed map, returning a Val
+func (ak ArgAnnotationKey) Lookup(annMap Map) (Val, bool) {
+	if paramVal, ok := annMap.CheckFuncParamAnn(ak.FuncDecl, ak.ParamNum); ok {
+		return paramVal, true
+	}
+	return nonAnnotatedDefault, false
+}
+
+// Object returns the types.Object that this annotation can best be interpreted as annotating
+func (ak ArgAnnotationKey) Object() types.Object {
+	return ak.FuncDecl
+}
+
+func (ak ArgAnnotationKey) String() string {
+	argname := ""
+	if ak.ParamName() != nil {
+		argname = fmt.Sprintf(": '%s'", ak.ParamName().Name())
+	}
+	return fmt.Sprintf("Param %d%s of Function %s at Location %s",
+		ak.ParamNum, argname, ak.FuncDecl.Name(), ak.Location.String())
+}
+
+// MinimalString returns a string representation for this ArgAnnotationKey consisting only
+// of the word "arg" followed by the name of the parameter, if named, or its position otherwise
+func (ak ArgAnnotationKey) MinimalString() string {
+	if ak.ParamName() != nil && len(ak.ParamName().Name()) > 0 {
+		return fmt.Sprintf("arg `%s`", ak.ParamName().Name())
+	}
+	return fmt.Sprintf("arg %d", ak.ParamNum)
```

Comment:
Correct. We don't have such a test case. The only contract format that could make sense for such a function with unnamed parameter is for example (_ -> ...) but again this contract does not say anything useful on the parameter.

I copied the code basically from the existing `MinimalString()` method for `ParamAnnotationKey`. We can 
1) Keep as is, although the else branch would be not reached I believe.
2) Remove the else branch and the if condition, assuming we always have a named parameter for a `CallSiteAnnotationKey`.

Personally I am inclined to 1 because it requires no assumption and is thus decoupled with any further changes we could make in other places.

Reply:
Yes, I agree. Even if we go for 2, we would basically have to add some sort of assertion/check and panic if we ever see such malformed contract, whereas this is more robust.

We could even have a test cases of the style:

```
// Contract below isn't useful, since return is always nonnil and argument
// is ignored, but added to check we don't crash on unnamed parameters
// contract(nonnil -> nonnil)
func foo(_ *int) *int {
  return &1
}
```

Reply:
Cool, added.